src/numberparse/correct.rs

```diff
@@ -0,0 +1,526 @@
+mod numberconst;
+use self::numberconst::{MANTISSA_128, POW10, POW10_COMPONENTS};
+use super::{is_integer, is_not_structural_or_whitespace_or_exponent_or_decimal};
+
+#[cfg(feature = "swar-number-parsing")]
+use super::{is_made_of_eight_digits_fast, parse_eight_digits_unrolled};
+
+use crate::charutils::is_structural_or_whitespace;
+use crate::error::Error;
+use crate::unlikely;
+use crate::StaticNode;
+use crate::{mem, static_cast_i64, Deserializer, ErrorType, Result};
+
+macro_rules! get {
+    ($buf:ident, $idx:expr) => {
+        unsafe { *$buf.get_unchecked($idx as usize) }
+    };
+}
+macro_rules! err {
+    ($idx:ident, $num:expr) => {
+        return Err(Error::new($idx, $num as char, ErrorType::InvalidNumber))
+    };
+}
+
+#[cfg_attr(not(feature = "no-inline"), inline)]
+#[allow(clippy::cast_possible_truncation)]
+fn multiply_as_u128(a: u64, b: u64) -> (u64, u64) {
+    let res: u128 = u128::from(a) * u128::from(b);
+    (res as u64, (res >> 64) as u64)
+}
+
+impl<'de> Deserializer<'de> {
+    #[allow(
+        unused_unsafe,
+        clippy::cast_possible_wrap,
+        clippy::cast_possible_truncation,
+        clippy::too_many_lines
+    )]
+    pub(crate) fn parse_number(start_idx: usize, buf: &[u8], negative: bool) -> Result<StaticNode> {
```

Comment:
can we use `idx` instead of `start_idx` the same here as we do in the approx implementation (or the other way around) the two implementations having the same parameter types with different meanings seems like it'll lead to headache down the road - it would also remove the need for the feature flags and code duplication in `stage2,rs`

Reply:
> it would also remove the need for the feature flags and code duplication in stage2,rs

I tried but I couldn't remove that duplication. I guess the new algorithm is not using the `get!` macro like the old one.

Anyway, I changed the signature to use `idx`. I had to kept the `start_idx` variable because it is used in the algo.

Reply:
[Got it](https://github.com/simd-lite/simd-json/pull/253/commits/fa8df70f5cd0c14388fce8fc8cef646db40531ef)! Thank you!